cpp/terrain/terrain_search.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include <cmath>
+
+#include "terrain.h"
+#include "terrain_object.h"
+#include "terrain_search.h"
+
+namespace openage {
+
+std::shared_ptr<TerrainObject> find_near(const TerrainObject &start, 
+                                         std::function<bool(const TerrainObject &)> found) {
+	
+	auto terrain = start.get_terrain();
+	auto tile = start.pos.draw.to_tile3().to_tile();
+	TerrainSearch search(terrain, tile);
+
+	for (int i = 0; i < 500; ++i) {
+		for (auto o : terrain->get_data(tile)->obj) {
+			auto shared_o = o.lock();
```

Comment:
Does `lock()` always succeed (can we be sure the `weak_ptr` has not expired?)

If yes, write a comment for stuff like that. If no, do a check (unless `found` must do the check... but the non-existing doc for `find_near` doesn't say so)